ui/src/storage/local.ts

```diff
@@ -0,0 +1,23 @@
+import * as tanagra from "tanagra-api";
+import { registerStoragePlugin, StoragePlugin } from "./storage";
+
+const storageKey = "tanagra-userdata";
```

Comment:
It might make sense to have the key instead be: `tanagra-userdata-${version}`. This would enable a user to continue using the app on the new version, even if we fail to migrate their old data version, without data loss. It does open up more complicated merge/migration scenarios, but it's probably better to be explicit in these cases.

This also would open up more options with rollbacks.

Reply:
Right.. the old app not being able to read a new version is exactly why the current approach makes rollbacks impossible; any user that visited the app (and had their localstorage migrated) between initial deployment and rollback would be broken. If this is intended as a stop-gap until we have proper backend storage, then this is fine. Else we will eventually need to consider how the model can support this, even if it is messier.

Reply:
The problem is that they're broken in the rollback case too. The older app can't correctly parse the newer data and trying to can potentially destroy their data. Or they get rolled back to the previous version of their data losing any progress in the new version and start making a new branch of edits it's not clear how we could resolve.

I'm also not sure what you mean by proper backend storage. Since the storage in most instances won't be controlled by us (i.e. we can't touch it except through the app), handling it in the backend is just moving the problem. It's actually even more complicated because then there's another potential version difference between the backend and the UI. Handling storage through the backend may still be correct for other reasons but it doesn't matter for this issue.

Let's discuss further and see what we can come up with but in my experience there aren't any rollback paths that are actually practical for end users.